### PR TITLE
continuously delete pods on nodes that don't exist

### LINKED ISSUE #22545: Daemonsets are forever
Two nodes went away as I updated the instances, but they still appear and, more importantly/confusingly, their daemonset pods are still in Running state. 

```
$ fleetctl list-machines
MACHINE         IP              METADATA
729947b9...     10.0.12.180     role=node
a0883347...     10.0.8.4        instance=m0,role=master
$ kubectl get nodes
NAME                          STATUS                     AGE
ip-10-0-10-222.ec2.internal   NotReady                   2d
ip-10-0-12-147.ec2.internal   NotReady                   1d
ip-10-0-12-180.ec2.internal   Ready                      17m
ip-10-0-8-4.ec2.internal      Ready,SchedulingDisabled   2d
$ kubectl get pods --all-namespaces
NAMESPACE     NAME                    READY     STATUS        RESTARTS   AGE
kube-system   fluentd-logging-2gxb1   1/1       Running       0          1d
kube-system   fluentd-logging-33y2f   1/1       Running       0          1d
kube-system   fluentd-logging-b1yuv   1/1       Running       0          17m
kube-system   fluentd-logging-o9hvp   1/1       Running       0          1d
```

There are only two instances and thus two daemons up, not four.
I understand the rationale for not killing the pods outright, but is there a timeout after which they change state or perhaps the nodes do?

My concern is twofold:
 1. `Running` is a lie. Things will be confusing for monitoring/accounting/troubleshooting purposes.
 1. State about pods and nodes is kept around for a long time. Imagine someone using DaemonSets on a pool of EC2 spot instances...

### PR DESCRIPTION
**Pull request by mikedanese**
Fixes #22545

### CODE REVIEW

File: pkg/controller/node/nodecontroller.go [hunk 1]
```diff
@@ -368,6 +371,29 @@ func (nc *NodeController) maybeDeleteTerminatingPod(obj interface{}) {
...
+	}
+
+	for _, pod := range pods {
+		if pod.Spec.NodeName == "" {
+			continue
+		}
+		if _, exists, _ := nc.nodeStore.Store.GetByKey(pod.Spec.NodeName); exists {
+			continue
+		}
+		fmt.Println("!!!!!!!!!!!!!!!!!!!!!!!!!!! %#v", nc.nodeStore.Store.List())
```
Comment (davidopp) [inline pkg/controller/node/nodecontroller.go]: I assume this was for debugging and should be removed or converted to a log message?

Comment (mikedanese) [inline pkg/controller/node/nodecontroller.go]: Fixed

**Comment by davidopp** [comment]
LGTM

File: pkg/controller/node/nodecontroller.go [hunk 2]
```diff
@@ -368,6 +371,28 @@ func (nc *NodeController) maybeDeleteTerminatingPod(obj interface{}) {
...
+	}
+
+	for _, pod := range pods {
+		if pod.Spec.NodeName == "" {
+			continue
+		}
+		if _, exists, _ := nc.nodeStore.Store.GetByKey(pod.Spec.NodeName); exists {
+			continue
+		}
+		if err := nc.forcefullyDeletePod(pod); err != nil {
```
Comment (mml) [inline pkg/controller/node/nodecontroller.go]: You will get here if GetByKey returns an error.  Also, is it possible for the cache to be stale?

Comment (mikedanese) [inline pkg/controller/node/nodecontroller.go]: GetByKey, can't return errors stupidly. That should be cleaned up:

https://github.com/kubernetes/kubernetes/blob/master/pkg/client/cache/store.go#L192

I'll do this in a follow up.

These cashes can be stale. There's a race here if a (node get's created -> scheduler get's update of node -> new pod gets assigned to a node -> controller get's update of pod) before (node get's created -> controller get's update of node). It seems less likely and of smaller consequence then the but this fixes. Any suggestion on how to work around this?

Comment (mikedanese) [inline pkg/controller/node/nodecontroller.go]: Also the node has to become ready in the scheduler flow and doesn't have to become ready in the controller manager flow which also makes the race less likely. Still a race though.

Comment (mml) [inline pkg/controller/node/nodecontroller.go]: My intuition is that this race is pretty minor and less significant than the bug it fixes, but our intuition is often wrong about this stuff, and on an antagonistically loaded machine, you could imagine being in some kind of steady state forever where the controllers fight each other.

But there's a non-controller behavior that's also new.  If I manually specify a host in my single pod (no controller), but make a typo, my pod disappears and never runs.  That'll seem weird.  It's also possible for the user to submit the pod in the same race window (hard to imagine this really happening) and the same thing happens.

Suggestions for fixing the race... I'll think about it.  Concurrency + global state + caching lends itself to this.

**Comment by gmarek** [comment]
This broke garbage collector tests: #22611

**Comment by wojtek-t** [comment]
I'm removing cherrypick-candidate label, since we are reverting this PR.

**Comment by wojtek-t** [comment]
@mikedanese @davidopp - I don't understand this PR at all. We have a GC controller that is supposed to (among others) delete such pods. Why we are adding such logic to node controller?

**Comment by davidopp** [comment]
The problem is that NodeController.deletePods() won't ever evict DaemonSet pods. So they never go to a terminated state when the user deletes the VM manually (which is what this bug is about). Because they never go to a terminated state, GC controller won't clean them up.

**Comment by wojtek-t** [comment]
But since they are orphaned, we are more talking about garbage collection here... So I think that we should put this logic to GC controller.

**Comment by davidopp** [comment]
Oh sorry, I thought you are asking "why do we need this logic" not "why did we put this logic in node controller."

Maybe not a good reason, but currently the GC controller doesn't have a nodeStore, it only thinks about pods... Whereas this operation requires both pods and nodes.

**Comment by wojtek-t** [comment]
>
Oh sorry, I thought you are asking "why do we need this logic" not "why did we put this logic in node controller."

Yeah - sorry. I wasn't clear.

**Comment by mikedanese** [comment]
Thanks for reverting, I'll fix the test and revert the revert.

**Comment by mikedanese** [comment]
Also this looks like a test bug not an issue with how the two loops interact.

**Pull request merged**